lua/nvim-treesitter/textobjects/select.lua

```diff
@@ -145,33 +146,52 @@ function M.detect_selection_mode(query_string, keymap_mode)
 end
 
 function M.attach(bufnr, lang)
-  local buf = bufnr or api.nvim_get_current_buf()
+  bufnr = bufnr or api.nvim_get_current_buf()
   local config = configs.get_module "textobjects.select"
-  lang = lang or parsers.get_buf_lang(buf)
+  lang = lang or parsers.get_buf_lang(bufnr)
 
   for mapping, query in pairs(config.keymaps) do
-    local desc
+    local desc, query_string, query_group
     if type(query) == "table" then
       desc = query.desc
-      query = query.query
+      query_string = query.query
+      query_group = query.query_group or "textobjects"
+    else
+      query_string = query
+      query_group = "textobjects"
     end
     if not desc then
-      desc = "Select textobject " .. query
+      desc = "Select textobject " .. query_string
     end
-    if not queries.get_query(lang, "textobjects") then
-      query = nil
+
+    local available_textobjects = shared.available_textobjects(lang, query_group)
+    local available = false
+    for _, available_textobject in ipairs(available_textobjects) do
+      if "@" .. available_textobject == query_string then
+        available = true
+        break
+      end
     end
-    if query then
-      local cmd_o = "<cmd>lua require'nvim-treesitter.textobjects.select'.select_textobject('" .. query .. "', 'o')<CR>"
-      local cmd_x = "<cmd>lua require'nvim-treesitter.textobjects.select'.select_textobject('" .. query .. "', 'x')<CR>"
-      vim.keymap.set({ "o" }, mapping, cmd_o, { buffer = buf, silent = true, remap = false, desc = desc })
-      vim.keymap.set({ "x" }, mapping, cmd_x, { buffer = buf, silent = true, remap = false, desc = desc })
+
+    if not available then
+      query_string = nil
+    end
+
+    if query_string then
+      local cmd_o = function()
+        M.select_textobject(query_string, query_group, "o")
+      end
+      local cmd_x = function()
+        M.select_textobject(query_string, query_group, "x")
```

Comment:
@kiyoon weren't lua functions causing problems in visual mode? There was a reason why this still was a string.

Reply:
@theHamsta I don't know, I think maybe before the patch from nvim-treesitter there were lots of weird bugs even with `<cmd>` and now it seems okay. It at least passes the tests with visual selection

Reply:
that's good. glad that we can finally get rid of this vimscript artifact.

Reply:
@theHamsta By the way, when does the `detach` function exactly called? I think my workflow is usually limited to having not many buffers at the same time, but it doesn't seem like it's being called whenever you close the buffer? Or I am doing something wrong